### ISSUE TITLE
Can we keep aggregation input cube lazy?

### ISSUE THREAD

**Issue by rcomer** [SciTools/iris#2928]
I already knew that `cube.aggregated_by` is not a lazy operation, but I just realized that the original cube's data becomes non-lazy after aggregation.  Example:

```python
import iris
import iris.coord_categorisation as cat

fname = iris.sample_data_path('SOI_Darwin.nc')
soi = iris.load_cube(fname)
cat.add_year(soi, 'time')
print soi.has_lazy_data()
soi.aggregated_by('year', iris.analysis.MEAN)
print soi.has_lazy_data()
```
gives:
```
True
False
```
I think* this is purely down to the statement `if ma.isMaskedArray(self.data):` at [line 3386](https://github.com/SciTools/iris/blob/master/lib/iris/cube.py#L3386).

I don't know anything about working with lazy arrays, but is there a way to check for masked/missing data without loading it?  If so, this appears to be easy to fix.

*I tried removing this if-loop in a copy of Iris 1.13, and the data stayed lazy.

**Comment by pelson** [comment]
Good question! Going back to to top - there is no reason we can't make aggregate_by lazy. I've spoken to a few people about doing precisely that, as well as parallelising it using the power of dask...

As for whether the cube being aggregated_by can remain lazy, we have to think about choosing a strategy that is consistent throughout iris, as there are a number of possible solutions:

 * original data is loaded when aggregated_by is called
 * original data is loaded when the aggregated_by result's data is realised
 * original data remains lazy, even when the aggregated_by result's data is realised

If we don't achieve the latter, it will not be possible to aggregate_by data >memory. If we do achieve the latter, the implication is that it will be easy to end up accidentally loading the original data more than once (especially if the aggregate_by operation is parallelised).

> I don't know anything about working with lazy arrays, but is there a way to check for masked/missing data without loading it

That is the primary difference between Iris' use of biggus vs dask: we used to be able to lazily ascertain whether data was masked or not, but dask prevents this from happening. We've worked around this in a few ways, but the most conspicuous is https://github.com/SciTools/iris/blob/master/lib/iris/fileformats/netcdf.py#L761, where we check the mask as it streams through the netcdf save function.

In summary, there is no reason why ``cube.aggregated_by`` should not become lazy - we just haven't done it yet.

xref: 2418

**Comment by rcomer** [comment]
Hi @pelson, thanks for your detailed explanation.  I think I’ll close this as it seems the answer to my specific question is “no” and the more general question of a lazy aggregated_by is covered elsewhere.
